client/client-rulesengine/src/main/java/software/amazon/smithy/java/client/rulesengine/EndpointUtils.java

```diff
@@ -0,0 +1,179 @@
+        if (value instanceof StringValue s) {
+            return s.getValue();
+        } else if (value instanceof IntegerValue i) {
+            return i.getValue();
+        } else if (value instanceof ArrayValue a) {
+            var result = new ArrayList<>();
+            for (var v : a.getValues()) {
+                result.add(convertInputParamValue(v));
+            }
+            return result;
+        } else if (value instanceof EmptyValue) {
+            return null;
+        } else if (value instanceof BooleanValue b) {
+            return b.getValue();
+        } else if (value instanceof RecordValue r) {
+            var result = new HashMap<>();
+            for (var e : r.getValue().entrySet()) {
+                result.put(e.getKey().getName().getValue(), convertInputParamValue(e.getValue()));
+            }
+            return result;
+        } else {
+            throw new RulesEvaluationError("Unsupported value type: " + value);
+        }
+    }
+
+    static Object verifyObject(Object value) {
+        if (value instanceof String
+                || value instanceof Number
+                || value instanceof Boolean
+                || value instanceof StringTemplate
+                || value instanceof URI) {
+            return value;
+        }
+
+        if (value instanceof List<?> l) {
+            for (var v : l) {
+                verifyObject(v);
+            }
+            return value;
+        }
+
+        if (value instanceof Map<?, ?> m) {
+            for (var e : m.entrySet()) {
+                if (!(e.getKey() instanceof String)) {
+                    throw new UnsupportedOperationException("Endpoint parameter maps must use string keys. Found " + e);
+                }
+                verifyObject(e.getKey());
+                verifyObject(e.getValue());
+            }
+            return m;
+        }
+
+        throw new UnsupportedOperationException("Unsupported endpoint rules value given: " + value);
+    }
+
+    // Read little-endian unsigned short (2 bytes)
+    static int bytesToShort(byte[] instructions, int offset) {
+        int low = instructions[offset] & 0xFF;
+        int high = instructions[offset + 1] & 0xFF;
+        return (high << 8) | low;
+    }
+
+    // Write little-endian unsigned short (2 bytes)
+    static void shortToTwoBytes(int value, byte[] instructions, int offset) {
+        instructions[offset] = (byte) (value & 0xFF);
+        instructions[offset + 1] = (byte) ((value >> 8) & 0xFF);
+    }
+
+    static Object getUriProperty(URI uri, String key) {
+        return switch (key) {
+            case "scheme" -> uri.getScheme();
+            case "path" -> uri.getRawPath();
+            case "normalizedPath" -> getUriNormalizedPath(uri);
+            case "authority" -> uri.getAuthority();
+            case "isIp" -> getUriIsIp(uri);
+            default -> null;
+        };
+    }
+
+    private static boolean getUriIsIp(URI uri) {
```

Comment:
For this method, see [this report](https://github.com/aws/aws-sdk-java-v2/issues/5846) about the impact of `Integer.parseInt` and how it was implemented in the Java SDK v2 to fix this issue [here](https://github.com/aws/aws-sdk-java-v2/blob/93f914840a071fb162693d1ebd753803281253cf/codegen/src/main/resources/software/amazon/awssdk/codegen/rules2/RuleUrl.java.resource#L58-L111)

Reply:
Thanks! I hope to address it by upstreaming to smithy/smithy: https://github.com/smithy-lang/smithy/pull/2617